integration-tests/src/tests/client/features/flat_storage.rs

```diff
@@ -0,0 +1,151 @@
+        receiver_id: "test0".parse().unwrap(),
+        public_key: signer.public_key(),
+        actions: vec![],
+        nonce: 0,
+        block_hash: CryptoHash::default(),
+    };
+
+    // Write key-value pair to state.
+    {
+        let write_value_action = vec![Action::FunctionCall(FunctionCallAction {
+            args: encode(&[1u64, 10u64]),
+            method_name: "write_key_value".to_string(),
+            gas,
+            deposit: 0,
+        })];
+        let tip = env.clients[0].chain.head().unwrap();
+        let signed_transaction = Transaction {
+            nonce: 10,
+            block_hash: tip.last_block_hash,
+            actions: write_value_action,
+            ..tx.clone()
+        }
+        .sign(&signer);
+        let tx_hash = signed_transaction.get_hash();
+        env.clients[0].process_tx(signed_transaction, false, false);
+        for i in 0..epoch_length / 3 {
+            env.produce_block(0, tip.height + i + 1);
+        }
+
+        let final_result = env.clients[0].chain.get_final_transaction_result(&tx_hash).unwrap();
+        assert_matches!(final_result.status, FinalExecutionStatus::SuccessValue(_));
+        let transaction_outcome = env.clients[0].chain.get_execution_outcome(&tx_hash).unwrap();
+        let receipt_ids = transaction_outcome.outcome_with_id.outcome.receipt_ids;
+        assert_eq!(receipt_ids.len(), 1);
+        let receipt_execution_outcome =
+            env.clients[0].chain.get_execution_outcome(&receipt_ids[0]).unwrap();
+        assert_matches!(
+            receipt_execution_outcome.outcome_with_id.outcome.status,
+            ExecutionStatus::SuccessValue(_)
+        );
+    }
+
+    let touching_trie_node_costs: Vec<_> = (0..2)
+        .map(|i| {
+            let read_value_action = vec![Action::FunctionCall(FunctionCallAction {
+                args: encode(&[1u64]),
+                method_name: "read_value".to_string(),
+                gas,
+                deposit: 0,
+            })];
+            let tip = env.clients[0].chain.head().unwrap();
+            let signed_transaction = Transaction {
+                nonce: 20 + i,
+                block_hash: tip.last_block_hash,
+                actions: read_value_action,
+                ..tx.clone()
+            }
+            .sign(&signer);
+            let tx_hash = signed_transaction.get_hash();
+            env.clients[0].process_tx(signed_transaction, false, false);
+            for i in 0..epoch_length / 3 {
+                env.produce_block(0, tip.height + i + 1);
+            }
+            if i == 0 {
+                env.upgrade_protocol(new_protocol_version);
+            }
+
+            let final_result = env.clients[0].chain.get_final_transaction_result(&tx_hash).unwrap();
+            assert_matches!(final_result.status, FinalExecutionStatus::SuccessValue(_));
+            let transaction_outcome = env.clients[0].chain.get_execution_outcome(&tx_hash).unwrap();
```

Comment:
Couldn't you use `transaction_outcome.transaction_outcome` instead of querying again?

Reply:
In this case I would get `ExecutionMetadataView` instead of `ExecutionMetadata`, and querying specific costs from there is ugly. However, I could reuse `receipt_id` from `final_transaction_result`.